log4j-core/src/main/java/org/apache/logging/log4j/core/appender/lucene/LuceneAppender.java

```diff
@@ -0,0 +1,295 @@
+	 * Delete the index before {@link LuceneAppender#expiryTime} second every day at 0.
+	 * 
+	 * @see LuceneAppender#expiryTime
+	 */
+	private void registerClearTimer() {
+		Calendar calendar = Calendar.getInstance();
+		long curMillis = calendar.getTimeInMillis();
+		calendar.add(Calendar.DAY_OF_MONTH, 1);
+		calendar.set(Calendar.HOUR_OF_DAY, 0);
+		calendar.set(Calendar.MINUTE, 0);
+		calendar.set(Calendar.SECOND, 0);
+		calendar.set(Calendar.MILLISECOND, 0);
+		long difMinutes = (calendar.getTimeInMillis() - curMillis) / (1000 * 60);
+
+		scheduledExecutor.scheduleAtFixedRate(new Runnable() {
+			@Override
+			public void run() {
+				LOGGER.info("delete index start!",target,expiryTime);
+				IndexWriter indexWriter = initIndexWriter();
+				if (null != indexWriter) {
+					Long start = 0L;
+					Long end = System.currentTimeMillis() - expiryTime * 1000;
+					NumericRangeQuery<Long> rangeQuery = NumericRangeQuery.newLongRange("timestamp", start, end, true, true);
+					try {
+						indexWriter.deleteDocuments(rangeQuery);
+						indexWriter.commit();
+						LOGGER.info("delete index end! ",target,expiryTime);
+					}catch (IOException e) {
+						LOGGER.error("delete index failed! "+e.getMessage(),e);
+					}
+				}
+			}
+		}, difMinutes, 1440, TimeUnit.MINUTES);
+	}
+
+	/**
+	 * create lucene index.
+	 * 
+	 * @param event
+	 */
+	@Override
+	public void append(LogEvent event) {
+		if (null != indexFields && indexFields.length > 0) {
+			IndexWriter indexWriter = initIndexWriter();
+			if(null != indexWriter){
+				Document doc = new Document();
+				doc.add(new LongField("timestamp", event.getTimeMillis(), Field.Store.YES));
+				try {
+					for (LuceneIndexField field : indexFields) {
+						String value = field.getLayout().toSerializable(event);
+						if (Strings.isEmpty(value) || value.matches("[$]\\{.+\\}")) {
+							return;
+						}
+						value = value.trim();
+						String type = field.getType();
+						if (Strings.isNotEmpty(type)) {
+							Class<?> clazz = Class.forName("org.apache.lucene.document." + type);
+							if (clazz == LongField.class) {
+								doc.add(new LongField(field.getName(), Long.valueOf(value), Field.Store.YES));
+							} else if (clazz == StringField.class) {
+								doc.add(new StringField(field.getName(), value, Field.Store.YES));
+							} else if (clazz == TextField.class) {
+								doc.add(new TextField(field.getName(), value, Field.Store.YES));
+							} else {
+								throw new UnsupportedOperationException(type + " type currently not supported.");
+							}
+						} else {
+							doc.add(new TextField(field.getName(), value, Field.Store.YES));
+						}
+					}
+					indexWriter.addDocument(doc);
+				} catch (Exception e) {
+					LOGGER.error(e.getMessage(), e);
+					if (!ignoreExceptions()) {
+						throw new AppenderLoggingException(e);
+					}
+				}
+			}
+		}
+	}
+
+	/**
+     * Factory method for creating a lucene appender within the plugin manager.
+     *
+     */
+	@PluginFactory

```

Comment:
Please use a builder instead of a factory method. This will allow us to avoid breaking binary compatibility in the future when new parameters are needed.